File: src/test/java/kitchenpos/menu/tobe/application/MenuServiceTest.java
package kitchenpos.menu.tobe.application;

import kitchenpos.common.exception.MenuException;
import kitchenpos.common.tobe.Profanities;
import kitchenpos.menu.tobe.domain.menu.*;
import kitchenpos.menu.tobe.domain.menugroup.MenuGroup;
import kitchenpos.menu.tobe.domain.menugroup.MenuGroupRepository;
import kitchenpos.menu.tobe.fake.FakePurogmalumClient;
import kitchenpos.menu.tobe.fake.InMemoryMenuGroupRepository;
import kitchenpos.menu.tobe.fake.InMemoryMenuRepository;
import kitchenpos.menu.tobe.fake.InMemoryProductRepository;
import kitchenpos.product.tobe.domain.Product;
import kitchenpos.product.tobe.domain.ProductRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class MenuServiceTest {

    private MenuService menuService;
    private MenuRepository menuRepository;
    private MenuGroupRepository menuGroupRepository;
    private ProductRepository productRepository;
    private Profanities profanities;
    private MenuValidator menuValidator;

    private UUID menuGroupId;
    private UUID productId1;
    private UUID productId2;
    private Product product1;
    private Product product2;

    @BeforeEach
    void setUp() {
        menuRepository = new InMemoryMenuRepository();
        menuGroupRepository = new InMemoryMenuGroupRepository();
        productRepository = new InMemoryProductRepository();
        profanities = new FakePurogmalumClient("바보", "멍청이");
        menuValidator = new MenuPriceValidator(productRepository);

        // 메뉴 그룹 생성
        menuGroupId = UUID.randomUUID();
        MenuGroup menuGroup = MenuGroup.of("음료");
        ReflectionTestUtils.setField(menuGroup, "id", menuGroupId);
        menuGroupRepository.save(menuGroup);

        // 상품 생성
        productId1 = UUID.randomUUID();
        productId2 = UUID.randomUUID();
        product1 = new Product("상품1", 5000L, profanities);
        product2 = new Product("상품2", 8000L, profanities);
        ReflectionTestUtils.setField(product1, "id", productId1);
        ReflectionTestUtils.setField(product2, "id", productId2);
        productRepository.save(product1);
        productRepository.save(product2);

        // 서비스 생성
        menuService = new MenuService(menuRepository, menuGroupRepository, profanities, menuValidator);
    }

    @Test
    @DisplayName("메뉴를 생성할 수 있다")
    void create() {
        // given
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        MenuRequest request = new MenuRequest(
                "아메리카노",
                13000L,
                menuGroupId,
                menuProducts,
                true
        );

        // when
        Menu created = menuService.create(request);

        // then
        assertThat(created).isNotNull();
        assertThat(created.getName()).isEqualTo("아메리카노");
        assertThat(created.getMenuPrice()).isEqualTo(13000L);
        assertThat(created.getMenuGroupId()).isEqualTo(menuGroupId);
        assertThat(created.isDisplayed()).isTrue();
        assertThat(created.getMenuProducts()).hasSize(2);
    }

    @Test
    @DisplayName("존재하지 않는 메뉴 그룹으로 메뉴를 생성할 수 없다")
    void createWithNonExistingMenuGroup() {
        // given
        UUID nonExistingMenuGroupId = UUID.randomUUID();
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        MenuRequest request = new MenuRequest(
                "아메리카노",
                13000L,
                nonExistingMenuGroupId,
                menuProducts,
                true
        );

        // when & then
        assertThatThrownBy(() -> menuService.create(request))
                .isInstanceOf(MenuException.class);
    }

    @Test
    @DisplayName("메뉴 가격을 변경할 수 있다")
    void changePrice() {
        // given
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        Menu menu = Menu.of("아메리카노", 13000L, menuGroupId, menuProducts, true, profanities, menuValidator);
        Menu saved = menuRepository.save(menu);

        MenuRequest request = new MenuRequest();
        request.setPrice(13000L);

        // when
        Menu updated = menuService.changePrice(saved.getId(), request);

        // then
        assertThat(updated.getMenuPrice()).isEqualTo(13000L);
    }

    @Test
    @DisplayName("메뉴를 display 상태로 변경할 수 있다")
    void display() {
        // given
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        Menu menu = Menu.of("아메리카노", 13000L, menuGroupId, menuProducts, false, profanities, menuValidator);
        Menu saved = menuRepository.save(menu);

        // when
        Menu displayed = menuService.display(saved.getId());

        // then
        assertThat(displayed.isDisplayed()).isTrue();
    }

    @Test
    @DisplayName("메뉴를 hide 상태로 변경할 수 있다")
    void hide() {
        // given
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        Menu menu = Menu.of("아메리카노", 13000L, menuGroupId, menuProducts, true, profanities, menuValidator);
        Menu saved = menuRepository.save(menu);

        // when
        Menu hidden = menuService.hide(saved.getId());

        // then
        assertThat(hidden.isDisplayed()).isFalse();
    }

    @Test
    @DisplayName("모든 메뉴를 조회할 수 있다")
    void findAll() {
        // given
        MenuProduct menuProduct1 = new MenuProduct(null, 1, 5000L, productId1);
        MenuProduct menuProduct2 = new MenuProduct(null, 1, 8000L, productId2);
        List<MenuProduct> menuProducts = Arrays.asList(menuProduct1, menuProduct2);

        Menu menu1 = Menu.of("아메리카노", 13000L, menuGroupId, menuProducts, true, profanities, menuValidator);
        Menu menu2 = Menu.of("카페라떼", 13000L, menuGroupId, menuProducts, true, profanities, menuValidator);

        menuRepository.save(menu1);
        menuRepository.save(menu2);

        // when
        List<Menu> menus = menuService.findAll();

        // then
        assertThat(menus).hasSize(2);
    }
}

File: src/test/kotlin/kitchenpos/menu/tobe/application/MenuServiceTest.kt
package kitchenpos.menu.tobe.application

import java.math.BigDecimal
import kitchenpos.menu.tobe.application.dto.CreateMenuProductReq
import kitchenpos.menu.tobe.application.dto.CreateMenuReq
import kitchenpos.menu.tobe.domain.MenuAmountService
import kitchenpos.menu.tobe.domain.MenuDisplay
import kitchenpos.menu.tobe.domain.MenuGroup
import kitchenpos.menu.tobe.domain.MenuGroupRepository
import kitchenpos.menu.tobe.domain.MenuNamePolicy
import kitchenpos.menu.tobe.domain.MenuProducts
import kitchenpos.menu.tobe.domain.MenuRepository
import kitchenpos.menu.tobe.domain.ProductClient
import kitchenpos.menu.tobe.infra.DefaultProductClient
import kitchenpos.menu.tobe.infra.FakeMenuGroupRepository
import kitchenpos.menu.tobe.infra.FakeMenuRepository
import kitchenpos.product.tobe.application.dto.ChangeProductPriceReq
import kitchenpos.product.tobe.domain.Product
import kitchenpos.product.tobe.domain.ProductRepository
import kitchenpos.product.tobe.infra.FakeProductRepository
import kitchenpos.product.tobe.infra.FakeProfanities
import kitchenpos.utils.Fixtures
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll

class MenuServiceTest {
    private lateinit var menuService: MenuService

    private lateinit var menuRepository: MenuRepository
    private lateinit var menuGroupRepository: MenuGroupRepository
    private lateinit var productRepository: ProductRepository
    private lateinit var productClient: ProductClient
    private lateinit var menuAmountService: MenuAmountService

    private lateinit var menuGroup: MenuGroup
    private lateinit var product: Product

    @BeforeEach
    fun setUp() {
        menuRepository = FakeMenuRepository()
        menuGroupRepository = FakeMenuGroupRepository()
        productRepository = FakeProductRepository()
        productClient = DefaultProductClient(productRepository)
        menuAmountService = DefaultMenuAmountService(productClient)
        menuService =
            MenuService(
                menuRepository,
                menuGroupRepository,
                productRepository,
                MenuNamePolicy(FakeProfanities()),
                menuAmountService
            )

        menuGroup = menuGroupRepository.save(Fixtures.menuGroup(name = "추천메뉴"))
        product = productRepository.save(Fixtures.product(name = "후라이드", price = 16_000))
    }

    @DisplayName("Menu를 등록한다")
    @Test
    fun create() {
        // given
        val request = CreateMenuReq(
            menuGroupId = menuGroup.id!!,
            name = "후라이드1마리",
            price = 16_000,
            display = MenuDisplay.DISPLAYED,
            menuProducts = listOf(
                CreateMenuProductReq(
                    seq = 1,
                    productId = product.id!!,
                    quantity = 1
                )
            )
        )

        // when
        val menuResp = menuService.create(request)

        // then
        assertThat(menuResp).isNotNull
        assertAll(
            { assertThat(menuResp.id).isNotNull() },
            { assertThat(menuResp.name).isEqualTo("후라이드1마리") },
            { assertThat(menuResp.price).isEqualTo(BigDecimal.valueOf(16000)) },
            { assertThat(menuResp.display).isEqualTo(MenuDisplay.DISPLAYED) },
            { assertThat(menuResp.menuGroup.id).isEqualTo(menuGroup.id) },
            { assertThat(menuResp.menuProducts).hasSize(1) }
        )
    }

    @DisplayName("Menu를 등록할 때 MenuProduct의 Product가 존재하지 않는다")
    @Test
    fun createWithInvalidProducts() {
        // given
        val request = CreateMenuReq(
            menuGroupId = menuGroup.id!!,
            name = "후라이드1마리",
            price = 16_000,
            display = MenuDisplay.DISPLAYED,
            menuProducts = listOf(
                CreateMenuProductReq(
                    seq = 1,
                    productId = product.id!!,
                    quantity = 1
                ),
                CreateMenuProductReq(
                    seq = 2,
                    productId = Fixtures.INVALID_UUID,
                    quantity = 1
                )
            )
        )

        // when then
        assertThatThrownBy { menuService.create(request) }
            .isInstanceOf(NoSuchElementException::class.java)
    }

    @DisplayName("Menu를 등록할 때 MenuGroup이 존재하지 않는다")
    @Test
    fun createWithInvalidMenuGroup() {
        // given
        val request = CreateMenuReq(
            menuGroupId = Fixtures.INVALID_UUID,
            name = "후라이드1마리",
            price = 16_000,
            display = MenuDisplay.DISPLAYED,
            menuProducts = listOf(
                CreateMenuProductReq(
                    seq = 1,
                    productId = product.id!!,
                    quantity = 1
                )
            )
        )

        // when then
        assertThatThrownBy { menuService.create(request) }
            .isInstanceOf(NoSuchElementException::class.java)
    }


    @DisplayName("Menu를 등록하려면 MenuPrice <= MenuAmount를 만족해야한다")
    @Test
    fun createWithInvalidPrice() {
        // given
        val request = CreateMenuReq(
            menuGroupId = menuGroup.id!!,
            name = "후라이드2마리",
            price = 32_001,
            display = MenuDisplay.DISPLAYED,
            menuProducts = listOf(
                CreateMenuProductReq(
                    seq = 1,
                    productId = product.id!!,
                    quantity = 2
                )
            )
        )

        // when then
        assertThatIllegalArgumentException().isThrownBy { menuService.create(request) }
    }

    @DisplayName("MenuPrice를 변경한다")
    @Test
    fun changePrice() {
        // given
        val menu = menuRepository.save(
            Fixtures.menu(
                menuAmountService = menuAmountService,
                name = "후라이드2마리",
                price = 32_000,
                display = MenuDisplay.DISPLAYED,
                menuGroup = menuGroup,
                menuProducts = MenuProducts(
                    listOf(Fixtures.menuProduct(productId = product.id!!, quantity = 2))
                ),
            )
        )
        val request = ChangeProductPriceReq(
            price = BigDecimal.valueOf(31_000)
        )

        // when
        menuService.changePrice(menu.id!!, request)

        // then
        assertThat(menu.menuPrice.price).isEqualTo(BigDecimal.valueOf(31_000))
    }

    @DisplayName("존재하지 않는 Menu는 MenuPrice를 변경할 수 없다")
    @Test
    fun changePriceInvalidMenuId() {
        // given
        val request = ChangeProductPriceReq(
            price = BigDecimal.valueOf(31_000)
        )

        // when then
        assertThatThrownBy {
            menuService.changePrice(
                Fixtures.INVALID_UUID,
                request
            )
        }.isInstanceOf(NoSuchElementException::class.java)
    }

    @DisplayName("MenuPrice를 변경할 때 MenuPrice <= MenuAmount를 만족해야한다")
    @Test
    fun changePriceLessThanAmount() {
        // given
        val menu = menuRepository.save(
            Fixtures.menu(
                menuAmountService = menuAmountService,
                name = "후라이드2마리",
                price = 32_000,
                display = MenuDisplay.DISPLAYED,
                menuGroup = menuGroup,
                menuProducts = MenuProducts(
                    listOf(Fixtures.menuProduct(productId = product.id!!, quantity = 2))
                ),
            )
        )
        val request = ChangeProductPriceReq(
            price = BigDecimal.valueOf(33_000)
        )

        // when then
        assertThatIllegalArgumentException().isThrownBy {
            menuService.changePrice(menu.id!!, request)
        }
    }

    @DisplayName("Menu를 Display한다")
    @Test
    fun menuDisplay() {
        // given
        val menu = menuRepository.save(
            Fixtures.menu(
                menuAmountService = menuAmountService,
                name = "후라이드2마리",
                price = 32_000,
                display = MenuDisplay.NOT_DISPLAYED,
                menuGroup = menuGroup,
                menuProducts = MenuProducts(
                    listOf(Fixtures.menuProduct(productId = product.id!!, quantity = 2))
                ),
            )
        )

        // when
        menuService.display(menu.id!!)

        // then
        assertThat(menu.menuDisplay).isEqualTo(MenuDisplay.DISPLAYED)
    }

    @DisplayName("존재하지 않는 Menu는 Displayed할 수 없다")
    @Test
    fun displayInvalidMenuId() {
        // when then
        assertThatThrownBy {
            menuService.display(Fixtures.INVALID_UUID)
        }.isInstanceOf(NoSuchElementException::class.java)
    }

    @DisplayName("Menu를 Not Displayed 한다")
    @Test
    fun menuNotDisplay() {
        // given
        val menu = menuRepository.save(
            Fixtures.menu(
                menuAmountService = menuAmountService,
                name = "후라이드2마리",
                price = 32_000,
                display = MenuDisplay.DISPLAYED,
                menuGroup = menuGroup,
                menuProducts = MenuProducts(
                    listOf(Fixtures.menuProduct(productId = product.id!!, quantity = 2))
                ),
            )
        )

        // when
        menuService.notDisplay(menu.id!!)

        // then
        assertThat(menu.menuDisplay).isEqualTo(MenuDisplay.NOT_DISPLAYED)
    }

    @DisplayName("존재하지 않는 Menu는 Not Displayed 할 수 없다")
    @Test
    fun notDisplayInvalidMenuId() {
        // when then
        assertThatThrownBy {
            menuService.notDisplay(Fixtures.INVALID_UUID)
        }.isInstanceOf(NoSuchElementException::class.java)
    }

}
